segmentation violation when running trivy in convert mode

### Discussed in https://github.com/aquasecurity/trivy/discussions/6775

<div type='discussions-op-text'>

<sup>Originally posted by **scott-boost** May 25, 2024</sup>
### Description

When I try to convert a Trivy JSON to CycloneDX, I get the following error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x28 pc=0x10941ea6c]
```

### Desired Behavior

a cyclonedx json file

### Actual Behavior

segmentation violation error

### Reproduction Steps

```bash
1. `trivy image --format cyclonedx ubuntu:latest --output /tmp/cdx_without_vulns.json`
2. `trivy sbom --format json --output /tmp/trivy_with_vulns.json /tmp/cdx_without_vulns.json`
3. `trivy convert --format cyclonedx  /tmp/trivy_with_vulns.json`
```


### Target

SBOM

### Scanner

Vulnerability

### Output Format

CycloneDX

### Mode

Standalone

### Debug Output

```bash
2024-05-24T14:06:54-04:00       DEBUG   ["cyclonedx" "spdx" "spdx-json" "github"] automatically enables '--list-all-pkgs'.
2024-05-24T14:06:54-04:00       DEBUG   Parsed severities       severities=[UNKNOWN LOW MEDIUM HIGH CRITICAL]
2024-05-24T14:06:54-04:00       INFO    "--format cyclonedx" disables security scanning. Specify "--scanners vuln" explicitly if you want to include vulnerabilities in the CycloneDX report.
2024-05-24T14:06:54-04:00       DEBUG   Writing report to output...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x28 pc=0x106772a6c]

goroutine 1 [running]:
github.com/aquasecurity/trivy/pkg/sbom/core.(*BOM).Root(0x0)
        github.com/aquasecurity/trivy/pkg/sbom/core/bom.go:279 +0x1c
github.com/aquasecurity/trivy/pkg/sbom/io.(*Encoder).rootComponent(_, {0x2, {0x137f2660, 0xedde2cb7d, 0x10f5ce840}, {0x14001720ca0, 0x1b}, {0x14002cdc9a7, 0x9}, {0x0, ...}, ...})
        github.com/aquasecurity/trivy/pkg/sbom/io/encode.go:86 +0x140
github.com/aquasecurity/trivy/pkg/sbom/io.(*Encoder).Encode(_, {0x2, {0x137f2660, 0xedde2cb7d, 0x10f5ce840}, {0x14001720ca0, 0x1b}, {0x14002cdc9a7, 0x9}, {0x0, ...}, ...})
        github.com/aquasecurity/trivy/pkg/sbom/io/encode.go:31 +0x4c
github.com/aquasecurity/trivy/pkg/sbom/cyclonedx.(*Marshaler).MarshalReport(_, {_, _}, {0x2, {0x137f2660, 0xedde2cb7d, 0x10f5ce840}, {0x14001720ca0, 0x1b}, {0x14002cdc9a7, ...}, ...})
        github.com/aquasecurity/trivy/pkg/sbom/cyclonedx/marshal.go:52 +0x6c
github.com/aquasecurity/trivy/pkg/report/cyclonedx.Writer.Write({{_, _}, _, {{_, _}, _, _}}, {_, _}, {0x2, ...})
        github.com/aquasecurity/trivy/pkg/report/cyclonedx/cyclonedx.go:31 +0x78
github.com/aquasecurity/trivy/pkg/report.Write({_, _}, {0x2, {0x137f2660, 0xedde2cb7d, 0x10f5ce840}, {0x14001720ca0, 0x1b}, {0x14002cdc9a7, 0x9}, ...}, ...)
        github.com/aquasecurity/trivy/pkg/report/writer.go:99 +0x778
github.com/aquasecurity/trivy/pkg/commands/convert.Run({_, _}, {{{0x108856773, 0xa}, 0x0, 0x0, 0x1, 0x0, 0x45d964b800, {0x1400281a8a0, ...}, ...}, ...})
        github.com/aquasecurity/trivy/pkg/commands/convert/run.go:43 +0x498
github.com/aquasecurity/trivy/pkg/commands.NewConvertCommand.func2(0x140028eb208, {0x14002b6e840, 0x1, 0x4})
        github.com/aquasecurity/trivy/pkg/commands/app.go:525 +0x154
github.com/spf13/cobra.(*Command).execute(0x140028eb208, {0x14002b6e800, 0x4, 0x4})
        github.com/spf13/cobra@v1.8.0/command.go:983 +0x840
github.com/spf13/cobra.(*Command).ExecuteC(0x14000205208)
        github.com/spf13/cobra@v1.8.0/command.go:1115 +0x344
github.com/spf13/cobra.(*Command).Execute(0x1088c4026?)
        github.com/spf13/cobra@v1.8.0/command.go:1039 +0x1c
main.run()
        github.com/aquasecurity/trivy/cmd/trivy/main.go:41 +0x158
main.main()
        github.com/aquasecurity/trivy/cmd/trivy/main.go:19 +0x20
```


### Operating System

macOS Sonoma 14.4.1

### Version

```bash
Version: 0.51.4
```


### Checklist

- [X] Run `trivy image --reset`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)</div>

--- a/pkg/sbom/io/encode.go
+++ b/pkg/sbom/io/encode.go
@@ -83,8 +83,15 @@ func (e *Encoder) rootComponent(r types.Report) (*core.Component, error) {
 		root.Type = core.TypeFilesystem
 	case artifact.TypeRepository:
 		root.Type = core.TypeRepository
-	case artifact.TypeCycloneDX:
-		return r.BOM.Root(), nil
+	case artifact.TypeCycloneDX, artifact.TypeSPDX:
+		// When we scan SBOM file
+		if r.BOM != nil {
+			return r.BOM.Root(), nil
+		}
+		// When we scan a `json` file (meaning a file in `json` format) which was created from the SBOM file.
+		// e.g. for use in `convert` mode.
+		// See https://github.com/aquasecurity/trivy/issues/6780
+		root.Type = core.TypeFilesystem
 	}
 
 	if r.Metadata.Size != 0 {

--- a/pkg/sbom/io/encode_test.go
+++ b/pkg/sbom/io/encode_test.go
@@ -20,7 +20,6 @@ func TestEncoder_Encode(t *testing.T) {
 	tests := []struct {
 		name           string
 		report         types.Report
-		addBOMFunc     func() *core.BOM
 		wantComponents map[uuid.UUID]*core.Component
 		wantRels       map[uuid.UUID][]core.Relationship
 		wantVulns      map[uuid.UUID][]core.Vulnerability
@@ -537,8 +536,7 @@ func TestEncoder_Encode(t *testing.T) {
 			wantVulns: make(map[uuid.UUID][]core.Vulnerability),
 		},
 		{
-			name:       "SBOM file",
-			addBOMFunc: newTestBOM,
+			name: "SBOM file",
 			report: types.Report{
 				SchemaVersion: 2,
 				ArtifactName:  "report.cdx.json",
@@ -566,19 +564,20 @@ func TestEncoder_Encode(t *testing.T) {
 						},
 					},
 				},
+				BOM: newTestBOM(t),
 			},
 			wantComponents: map[uuid.UUID]*core.Component{
-				uuid.MustParse("3ff14136-e09f-4df9-80ea-000000000001"): appComponent,
-				uuid.MustParse("3ff14136-e09f-4df9-80ea-000000000002"): libComponent,
+				uuid.MustParse("2ff14136-e09f-4df9-80ea-000000000001"): appComponent,
+				uuid.MustParse("3ff14136-e09f-4df9-80ea-000000000001"): libComponent,
 			},
 			wantRels: map[uuid.UUID][]core.Relationship{
-				uuid.MustParse("3ff14136-e09f-4df9-80ea-000000000001"): {
+				uuid.MustParse("2ff14136-e09f-4df9-80ea-000000000001"): {
 					{
-						Dependency: uuid.MustParse("3ff14136-e09f-4df9-80ea-000000000002"),
+						Dependency: uuid.MustParse("3ff14136-e09f-4df9-80ea-000000000001"),
 						Type:       core.RelationshipContains,
 					},
 				},
-				uuid.MustParse("3ff14136-e09f-4df9-80ea-000000000002"): nil,
+				uuid.MustParse("3ff14136-e09f-4df9-80ea-000000000001"): nil,
 			},
 			wantVulns: make(map[uuid.UUID][]core.Vulnerability),
 		},
@@ -654,9 +653,6 @@ func TestEncoder_Encode(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 			uuid.SetFakeUUID(t, "3ff14136-e09f-4df9-80ea-%012d")
 
-			if tt.addBOMFunc != nil {
-				tt.report.BOM = tt.addBOMFunc()
-			}
 			opts := core.Options{GenerateBOMRef: true}
 			got, err := sbomio.NewEncoder(opts).Encode(tt.report)
 			if tt.wantErr != "" {
@@ -732,7 +728,8 @@ var (
 	}
 )
 
-func newTestBOM() *core.BOM {
+func newTestBOM(t *testing.T) *core.BOM {
+	uuid.SetFakeUUID(t, "2ff14136-e09f-4df9-80ea-%012d")
 	bom := core.NewBOM(core.Options{})
 	bom.AddComponent(appComponent)
 	return bom

--- a/pkg/sbom/io/encode_test.go
+++ b/pkg/sbom/io/encode_test.go
@@ -20,6 +20,7 @@ func TestEncoder_Encode(t *testing.T) {
 	tests := []struct {
 		name           string
 		report         types.Report
+		addBOMFunc     func() *core.BOM
 		wantComponents map[uuid.UUID]*core.Component
 		wantRels       map[uuid.UUID][]core.Relationship
 		wantVulns      map[uuid.UUID][]core.Vulnerability
@@ -536,6 +537,97 @@ func TestEncoder_Encode(t *testing.T) {
 			wantVulns: make(map[uuid.UUID][]core.Vulnerability),
 		},
 		{
+			name:       "SBOM file",
+			addBOMFunc: newTestBOM,
+			report: types.Report{
+				SchemaVersion: 2,
+				ArtifactName:  "report.cdx.json",
+				ArtifactType:  artifact.TypeCycloneDX,
+				Results: []types.Result{
+					{
+						Target: "Java",
+						Type:   ftypes.Jar,
+						Class:  types.ClassLangPkg,
+						Packages: []ftypes.Package{
+							{
+								ID:      "org.apache.logging.log4j:log4j-core:2.23.1",
+								Name:    "org.apache.logging.log4j:log4j-core",
+								Version: "2.23.1",
+								Identifier: ftypes.PkgIdentifier{
+									PURL: &packageurl.PackageURL{
+										Type:      packageurl.TypeMaven,
+										Namespace: "org.apache.logging.log4j",
+										Name:      "log4j-core",
+										Version:   "2.23.1",
+									},
+								},
+								FilePath: "log4j-core-2.23.1.jar",
+							},
+						},
+					},
+				},
+			},
+			wantComponents: map[uuid.UUID]*core.Component{
+				uuid.MustParse("3ff14136-e09f-4df9-80ea-000000000001"): appComponent,
+				uuid.MustParse("3ff14136-e09f-4df9-80ea-000000000002"): libComponent,
+			},
+			wantRels: map[uuid.UUID][]core.Relationship{
+				uuid.MustParse("3ff14136-e09f-4df9-80ea-000000000001"): {
+					{
+						Dependency: uuid.MustParse("3ff14136-e09f-4df9-80ea-000000000002"),
+						Type:       core.RelationshipContains,
+					},
+				},
+				uuid.MustParse("3ff14136-e09f-4df9-80ea-000000000002"): nil,
+			},
+			wantVulns: make(map[uuid.UUID][]core.Vulnerability),
+		},
+		{
+			name: "json file created from SBOM file (BOM is empty)",
+			report: types.Report{
+				SchemaVersion: 2,
+				ArtifactName:  "report.cdx.json",
+				ArtifactType:  artifact.TypeCycloneDX,
+				Results: []types.Result{
+					{
+						Target: "Java",
+						Type:   ftypes.Jar,
+						Class:  types.ClassLangPkg,
+						Packages: []ftypes.Package{
+							{
+								ID:      "org.apache.logging.log4j:log4j-core:2.23.1",
+								Name:    "org.apache.logging.log4j:log4j-core",
+								Version: "2.23.1",
+								Identifier: ftypes.PkgIdentifier{
+									PURL: &packageurl.PackageURL{
+										Type:      packageurl.TypeMaven,
+										Namespace: "org.apache.logging.log4j",
+										Name:      "log4j-core",
+										Version:   "2.23.1",
+									},
+								},
+								FilePath: "log4j-core-2.23.1.jar",
+							},
+						},
+					},
+				},
+			},
+			wantComponents: map[uuid.UUID]*core.Component{
+				uuid.MustParse("3ff14136-e09f-4df9-80ea-000000000001"): fsComponent,
+				uuid.MustParse("3ff14136-e09f-4df9-80ea-000000000002"): libComponent,
+			},
+			wantRels: map[uuid.UUID][]core.Relationship{
+				uuid.MustParse("3ff14136-e09f-4df9-80ea-000000000001"): {
+					{
+						Dependency: uuid.MustParse("3ff14136-e09f-4df9-80ea-000000000002"),
+						Type:       core.RelationshipContains,
+					},
+				},
+				uuid.MustParse("3ff14136-e09f-4df9-80ea-000000000002"): nil,
+			},
+			wantVulns: make(map[uuid.UUID][]core.Vulnerability),
+		},
+		{
 			name: "invalid digest",
 			report: types.Report{
 				SchemaVersion: 2,
@@ -562,6 +654,9 @@ func TestEncoder_Encode(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 			uuid.SetFakeUUID(t, "3ff14136-e09f-4df9-80ea-%012d")
 
+			if tt.addBOMFunc != nil {
+				tt.report.BOM = tt.addBOMFunc()
+			}
 			opts := core.Options{GenerateBOMRef: true}
 			got, err := sbomio.NewEncoder(opts).Encode(tt.report)
 			if tt.wantErr != "" {
@@ -579,4 +674,66 @@ func TestEncoder_Encode(t *testing.T) {
 			assert.Equal(t, tt.wantVulns, got.Vulnerabilities())
 		})
 	}
+}
+
+var (
+	appComponent = &core.Component{
+		Root: true,
+		Type: core.TypeApplication,
+		Name: "log4j-core-2.23.1.jar",
+	}
+	fsComponent = &core.Component{
+		Root: true,
+		Type: core.TypeFilesystem,
+		Name: "report.cdx.json",
+		PkgIdentifier: ftypes.PkgIdentifier{
+			BOMRef: "3ff14136-e09f-4df9-80ea-000000000001",
+		},
+		Properties: core.Properties{
+			{
+				Name:  "SchemaVersion",
+				Value: "2",
+			},
+		},
+	}
+	libComponent = &core.Component{
+		Type:    core.TypeLibrary,
+		Name:    "log4j-core",
+		Group:   "org.apache.logging.log4j",
+		Version: "2.23.1",
+		PkgIdentifier: ftypes.PkgIdentifier{
+			BOMRef: "pkg:maven/org.apache.logging.log4j/log4j-core@2.23.1",
+			PURL: &packageurl.PackageURL{
+				Type:      packageurl.TypeMaven,
+				Namespace: "org.apache.logging.log4j",
+				Name:      "log4j-core",
+				Version:   "2.23.1",
+			},
+		},
+		Files: []core.File{
+			{
+				Path: "log4j-core-2.23.1.jar",
+			},
+		},
+		Properties: core.Properties{
+			{
+				Name:  "FilePath",
+				Value: "log4j-core-2.23.1.jar",
+			},
+			{
+				Name:  "PkgID",
+				Value: "org.apache.logging.log4j:log4j-core:2.23.1",
+			},
+			{
+				Name:  "PkgType",
+				Value: "jar",
+			},
+		},
+	}
+)
+
+func newTestBOM() *core.BOM {
+	bom := core.NewBOM(core.Options{})
+	bom.AddComponent(appComponent)
+	return bom
 }

